libbeat/publisher/queue/diskqueue/acks_test.go

```diff
@@ -0,0 +1,302 @@
+					},
+					frameID(3),
+					&queuePosition{0, segmentHeaderSize + 225, 3},
+					nil,
+				},
+				{
+					"Acknowledge the first frame of segment 1",
+					[]*readFrame{
+						rf(1, 3, false, 100),
+					},
+					frameID(7),
+					&queuePosition{2, segmentHeaderSize + 100, 1},
+					segmentIDRef(1),
+				},
+				{
+					"Acknowledge the middle frame of segment 2",
+					[]*readFrame{
+						rf(2, 7, false, 100),
+					},
+					frameID(9),
+					&queuePosition{2, segmentHeaderSize + 300, 3},
+					nil,
+				},
+			},
+		},
+		{
+			name: "ACKing multiple segments only sends the final one to the core loop",
+			steps: []addFramesTestStep{
+				{
+					"Add four one-frame segments",
+					[]*readFrame{
+						rf(3, 3, true, 100),
+						rf(2, 2, true, 100),
+						rf(1, 1, true, 100),
+						rf(0, 0, true, 100),
+					},
+					frameID(4),
+					&queuePosition{3, segmentHeaderSize + 100, 1},
+					// We advanced from segment 0 to segment 3, so we expect
+					// segmentID 2 on the ACK channel.
+					segmentIDRef(2),
+				},
+			},
+		},
+		{
+			name:     "The first frame of a segment resets nextPosition.byteIndex",
+			frameID:  35,
+			position: queuePosition{9, 1000, 10},
+			steps: []addFramesTestStep{
+				{
+					"Add the beginning of segment 10 as the next frame",
+					[]*readFrame{
+						rf(10, 35, true, 100),
+					},
+					frameID(36),
+					&queuePosition{10, segmentHeaderSize + 100, 1},
+					// We advanced to segment 10, so we expect segmentID 9 on
+					// the ACK channel.
+					segmentIDRef(9),
+				},
+			},
+		},
+		{
+			// Usually on the first frame of a segment, nextPosition is updated
+			// to point to the beginning of the new segment. On the very first
+			// frame of a new run, we don't do this, to allow the position to be
+			// restored from a previous run in case we shut down partway through
+			// a segment.
+			name:     "The first frame after the queue opens doesn't overwrite nextPosition.byteIndex",
+			frameID:  0,
+			position: queuePosition{5, 1000, 10},
+			steps: []addFramesTestStep{
+				{
+					"Add the beginning of segment 10 as the next frame",
+					[]*readFrame{
+						rf(5, 0, true, 100),
+					},
+					frameID(1),
+					// The new 100-byte frame should just be added to the existing
+					// byte position.
+					&queuePosition{5, 1100, 11},
+					nil,
+				},
+			},
+		},
+		{
+			name: "Segments with old schema versions have the correct positions",
+			steps: []addFramesTestStep{
+				{
+					"Add a frame for a schema-0 segment",
+					[]*readFrame{
+						{
+							segment: &queueSegment{
+								schemaVersion: uint32Ref(0),
+							},
+							bytesOnDisk: 100,
+						},
+					},
+					frameID(1),
+					// Schema-0 segments had 4-byte headers
+					&queuePosition{0, 4 + 100, 1},
+					nil,
+				},
+			},
+		},
+	}
+
+	path := filepath.Join(dir, "state.dat")
+	stateFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer stateFile.Close()
```

Comment:
why share the state file between tests? Does that create a dependency between tests? Consider to have a separate directory per test, so states and contents is completely independent.

Reply:
Done